README.md
# PegasusHelper
The pegasus helper is a small helper plugin for ILIAS which is required to
operate the ILIAS Pegasus mobile application.

Features:
- Custom app login flow
- Download files from ILIAS except FileObjects
- Directly open ILIAS pages with SSO tokens
- Open personal news of the user
- Configure required REST plugin routes and client
- Configure dynamic theme of the ILIAS-Pegasus community app
- Basic user token statistic
- Basic plugin setup tests which verify your local ILIAS configuration



## Requirements
* Version: ILIAS 5.3 to 6
* PHP 5.6 or PHP 7.0 - 7.4

## Installation

### 1. Install RESTPlugin
Branch feature/sr-app-routes  
Start at your ILIAS root directory 
```bash
mkdir -p Customizing/global/plugins/Services/UIComponent/UserInterfaceHook/  
cd Customizing/global/plugins/Services/UIComponent/UserInterfaceHook/  
git clone https://github.com/studer-raimann/Ilias.RESTPlugin REST
cd REST
git checkout feature/sr-app-routes
```  
Update and activate the plugin in the ILIAS Plugin Administration.

### 2. Install PegausHelper
Start at your ILIAS root directory 
```bash
cd Customizing/global/plugins/Services/UIComponent/UserInterfaceHook/  
git clone https://github.com/studer-raimann/PegasusHelper.git  
```
Update and activate the plugin in the ILIAS Plugin Administration.

### 3. Get the API Secret  
- Go to the ILIAS Plugin Administration  
- Choose the Action 'Configure' of the REST Plugin  
- Click the Button 'Start Administration Panel'  
- Click the Button 'Manage API-Keys and Authorization Schemes'  
- Click the Button 'Modify' at the row 'ilias_pegasus'  

## Testing
If the installation of the REST- or PegasusHelper-plugin as described above fails, the test-script in the directory 'testing' may provide useful information

Start at your ILIAS root directory
```bash
cd Customizing/global/plugins/Services/UIComponent/UserInterfaceHook/PegasusHelper/testing/
php run.php
```
The script prints out feedback from the tests and writes a log-file 'results.log' in 'PegasusHelper/testing/'

Also read through the paragraph 'Caveats' below

## Update

### 1. Update RESTPlugin
Branch feature/sr-app-routes
Start at your ILIAS root directory

```bash
cd Customizing/global/plugins/Services/UIComponent/UserInterfaceHook/REST
git pull
```
Update / activate the plugin in the ILIAS Plugin Administration.

### 2. Update PegasusHelper
Start at your ILIAS root directory 

```bash
cd Customizing/global/plugins/Services/UIComponent/UserInterfaceHook/PegasusHelper
git pull
```

Update / activate the plugin in the ILIAS Plugin Administration.

## Caveats
### ILIAS http path

#### Description
If the host address of ILIAS is configured with http but requests to ILIAS are
redirected to https, the plugin migration of the PegasusHelper will fail.

The PegasusHelper configures the REST plugin while updating to ensure that all
routes are as expected by the Pegasus mobile application. In order to configure the REST
plugin the PegasusHelper adds all routes to the REST plugin with local http POST requests.

The redirect will transform the POST request to a GET request which is not understood by
the REST plugin which leads to the migration error of the PegasusHelper.

#### Solution
To ensure that no https redirects are done, the configuration in the ilias.ini.php has to
be adjusted as shown in the example below.

The ilias.ini.php is located in the root directory of the ILIAS installation.
```text
[server]
http_path = "https://your.ilias-installation.org"
```

## Versioning
We use SemVer for versioning. For the versions available, see the tags on this repository.

## License
This project is licensed under the GNU GPLv3 License - see the LICENSE.md file for details.

## Acknowledgments
[composer](https://getcomposer.org/)

## Contact

studer + raimann ag  
Farbweg 9  
3400 Burgdorf  
Switzerland

[info@studer-raimann.ch](mailto:info@studer-raimann.ch)  
<https://www.studer-raimann.ch>



CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [2.0.0]
### Fixed
- News link redirects are now working correctly on ILIAS 6
- Timeline links no longer result in a 404 error on ILIAS 6

### Deprecated
- PHP 5.x support
- ILIAS 5.3 support

### Removed
- ILIAS 5.2 support

## [1.1.5]
### Fixed
- Plugin max ILIAS version constraint

## [1.1.4]
### Added
- added token statistics (number of logins for past 30 90 180 days)
- configuration of REST client without requests to API

## [1.1.3]
### Added
- app theme icons

## [1.1.2]
### Fixed
- addition of table 'ui_uihk_pegasus_theme' 

## [1.1.1]
### Added
- external testing
### Changed
- modified feedback to user for failed tests (even if test fails, setup may still be ok)
### Fixed
- checking for presence of mysqli implementation
- installing and uninstalling the plugin

## [1.1.0]
### Added
- app theme coloring

## [1.0.2]
### Added
- testing for REST- and PegasusHelper-Plugins
- support for ILIAS 5.4

## [1.0.1]
### Fixed
- White page while opening ILIAS website with an auth token 

## [1.0.0]
### Added
- Changelog
- Download of documents with authentication token.
- Open personal news feed with authentication token.
### Changed
- Internal refactoring to speed up future releases.
### Fixed
- Fixed a bug which could lead to an access denied error with a valid authentication token.

## [0.0.11]
### Added
- Login
- Open links in ILIAS
- Open time line of a course in ILIAS


## [Unreleased]
### Added
### Changed
### Deprecated
### Removed
### Fixed
### Security


diagnose_checks.php
<?php

function checkILIASVersion($ilias) {
    $pass = $ilias["version"][0] == 5 && ($ilias["version"][1] == 3 || $ilias["version"][1] == 4);
    $info = ($pass) ? NULL : "supported ILIAS-versions are 5.3 or 5.4\n";
    $mandatory = true;

    return array($pass, $info, $mandatory);
}

function checkRESTVersion($rest) {
    $pass = $rest["version"] == [1, 0, 1];
    $info = ($pass) ? NULL : "REST-version must be 1.0.1\n";
    $mandatory = true;

    return array($pass, $info, $mandatory);
}

function checkPegasusHelperVersion($pegasusHelper) {
    $pass = $pegasusHelper["version"] == [1, 7, 3];
    $info = ($pass) ? NULL : "PegasusHelper-version must be 1.7.3\n";
    $mandatory = true;

    return array($pass, $info, $mandatory);
}

function checkMinMaxVersion($ilias, $plugin) {
    $v = $ilias["version"];
    $vmin = $plugin["ilias_min_version"];
    $vmax = $plugin["ilias_max_version"];

    $pass = ($vmin <= $v) && ($v <= $vmax);
    $info = ($pass) ? NULL : "the version " . arrayVersionToStr($v) . " is not contained in " . arrayVersionToStr($vmin) . " and " . arrayVersionToStr($vmax) . "\n";
    $mandatory = true;

    return array($pass, $info, $mandatory);
}

function checkRedirectStatement($code) {
    $code = explode("\n", $code);
    $pass = false;
    for ($i = 0; $i < count($code) - 1; $i++) {
        if(strpos($code[$i], "[server]") === 0 && strpos($code[$i + 1], "http_path") === 0) {
            $pass = true;
            break;
        }
    }

    $info = ($pass) ? NULL : "if requests to ILIAS are redirected to https, then the file 'ilias.ini.php' in the root-directory must contain the following two lines\n[server]\nhttp_path = \"https://YOUR.ILIAS-INSTALLATION.ORG\"" . "\n";
    $mandatory = false;

    return array($pass, $info, $mandatory);
}